sssd_test_framework/utils/authentication.py

```diff
@@ -256,6 +270,91 @@ def password_expired(self, username: str, password: str, new_password: str) -> b
+        Call ``su - $username`` and authenticate the user with passkey.
+
+        :param username: User name
+        :type username: str
+        :param pin: Passkey PIN.
+        :type pin: str | int
+        :param device: Path to local umockdev device file.
+        :type device: str
+        :param ioctl: Path to local umockdev ioctl file.
+        :type ioctl: str
+        :param script: Path to local umockdev script file
+        :type script: str
+        :return: Generated passkey mapping string.
+        :rtype: str
+        :return: True if authentication was successful, False otherwise.
+        :rtype: bool
+        """
+        self.fs.backup("/etc/sysconfig/sssd")
+        device_path = self.fs.upload_to_tmp(device, mode="a=r")
+        ioctl_path = self.fs.upload_to_tmp(ioctl, mode="a=r")
+        script_path = self.fs.upload_to_tmp(script, mode="a=r")
+
+        run_su = self.fs.mktmp(
+            rf"""
+            #!/bin/bash
+            set -ex
+            env | grep ^UMOCKDEV_ > /etc/sysconfig/sssd
+            printf "LD_PRELOAD=$LD_PRELOAD" >> /etc/sysconfig/sssd
+            systemctl restart sssd
+            chmod -R a+rwx $UMOCKDEV_DIR
+
+            su --shell /bin/sh nobody -c "su - '{username}'"
+            """,
+            mode="a=rx",
+        )
+
+        playback_umockdev = self.fs.mktmp(
+            rf"""
+            #!/bin/bash
+
+            LD_PRELOAD=/opt/random.so umockdev-run \
+                --device '{device_path}' \
+                --ioctl '/dev/hidraw1={ioctl_path}' \
+                --script '/dev/hidraw1={script_path}' \
+                -- '{run_su}'
+            """,
+            mode="a=rx",
+        )
+
+        result = self.host.ssh.expect(
+            rf"""
+            # It takes some time to get authentication failure
+            set timeout 10
+            set prompt "\n.*\[#\$>\] $"
+
+            spawn "{playback_umockdev}"
+
+            expect {{
+                "Insert your passkey device, then press ENTER*" {{send -- "\r"}}
+                timeout {{puts "expect result: Unexpected output"; exit 1}}
+                eof {{puts "expect result: Unexpected end of file"; exit 2}}
+            }}
+
+            expect {{
+                "Enter PIN:*" {{send -- "{pin}\r"}}
+                timeout {{puts "expect result: Unexpected output"; exit 1}}
+                eof {{puts "expect result: Unexpected end of file"; exit 2}}
+            }}
+
+            expect {{
+                -re $prompt {{puts "expect result: Password authentication successful"; exit 0}}
+                "Authentication failure" {{puts "expect result: Authentication failure"; exit 4}}
+                timeout {{puts "expect result: Unexpected output"; exit 1}}
+                eof {{puts "expect result: Unexpected end of file"; exit 2}}
+            }}
+
+            puts "expect result: Unexpected code path"
+            exit 3
+            """
+        )
+
+        return result.rc == 0
```

Comment:
can we return only result.rc only,
as while checking auth_deny test cases need to check result.rc = 4.

Reply:
Sure. Just can you explain why is this needed?

Reply:
To check the auth deny we need the return code as a 4.
So instead of returning result.rc == 0,
can we return only result.rc?

Reply:
We agreed with Madhuri to change it as:
```
if result.rc == 0:
    return True
elif result.rc == 4:
    return False
else:
    # error in expect script, we got unexpected text
    raise Expection(...)
```